server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/data/StarTreeDataWriter.java

```diff
@@ -0,0 +1,110 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.compositeindex.datacube.startree.fileformats.data;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.IndexOutput;
+import org.opensearch.index.compositeindex.datacube.startree.node.InMemoryTreeNode;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static org.opensearch.index.compositeindex.datacube.startree.node.FixedLengthStarTreeNode.SERIALIZABLE_DATA_SIZE_IN_BYTES;
+import static org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeUtils.ALL;
+
+/**
+ * Utility class for serializing a star-tree data structure.
+ *
+ * @opensearch.experimental
+ */
+public class StarTreeDataWriter {
+
+    private static final Logger logger = LogManager.getLogger(StarTreeDataWriter.class);
+
+    /**
+     * Writes the star-tree data structure.
+     *
+     * @param indexOutput the IndexOutput to write the star-tree data
+     * @param rootNode    the root node of the star-tree
+     * @param numNodes    the total number of nodes in the star-tree
+     * @param name        the name of the star-tree field
+     * @return the total size in bytes of the serialized star-tree data
+     * @throws IOException if an I/O error occurs while writing the star-tree data
+     */
+    public static long writeStarTree(IndexOutput indexOutput, InMemoryTreeNode rootNode, int numNodes, String name) throws IOException {
+        long totalSizeInBytes = (long) numNodes * SERIALIZABLE_DATA_SIZE_IN_BYTES;
+
+        logger.debug("Star tree data size in bytes : {} for star-tree field {}", totalSizeInBytes, name);
+
+        writeStarTreeNodes(indexOutput, rootNode);
+        return totalSizeInBytes;
+    }
+
+    /**
+     * Writes the star-tree nodes in a breadth-first order.
+     *
+     * @param output   the IndexOutput to write the nodes
+     * @param rootNode the root node of the star-tree
+     * @throws IOException if an I/O error occurs while writing the nodes
+     */
+    private static void writeStarTreeNodes(IndexOutput output, InMemoryTreeNode rootNode) throws IOException {
+        Queue<InMemoryTreeNode> queue = new LinkedList<>();
+        queue.add(rootNode);
+
+        int currentNodeId = 0;
+        while (!queue.isEmpty()) {
+            InMemoryTreeNode node = queue.remove();
+
+            if (node.children == null || node.children.isEmpty()) {
+                writeStarTreeNode(output, node, ALL, ALL);
+            } else {
+
+                // Sort all children nodes based on dimension value
+                List<InMemoryTreeNode> sortedChildren = new ArrayList<>(node.children.values());
+                sortedChildren.sort(
```

Comment:
Its not sorted as we add the children to hashMap. I tried one round of testing with LinkedHashMap , so far not facing issues.

Reply:
I have 50 tests failing because of this currently. Looks like mostly assertions based on the star tree documents. But I am not sure of the risk in breaking the flow. 
We are confident about the current approach. We also have binary search during queries dependent on top of sorted behaviour. I would be extremely cautious with this change.

Reply:
So basically the reason is star node separation. Once we those changes, all tests are passing.

We can pick up star node separation as part of this PR or next PR based on the complexity of it. When we include the change, we can convert this to LinkedHashMap and remove sorting.

I'll try to give confidence on switch to LinkedHashMap and sorting based on load testing with the changes - if we see failures or issues, lets not switch this logic this late otherwise we can switch.

Reply:
thanks @bharath-techie 
I have removed the Tree Node. Its almost 30 files now in the PR. We can take star node separation in the next PR where we can test it end to end with file formats.

What do you guys think?